fuzzers/run_fuzzer.py

```diff
@@ -0,0 +1,369 @@
+    """
+    p = subprocess.Popen(
+        "pstree -T {}".format(pid),
+        shell=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.DEVNULL,
+        stdin=subprocess.DEVNULL,
+        universal_newlines=True,
+    )
+    stdout, stderr = p.communicate()
+    return stdout
+
+
+def main(argv):
+    fuzzers = os.path.abspath(os.path.dirname(__file__))
+
+    assert len(argv) == 2, argv
+    fuzzer = argv[1]
+    time_start = datetime.utcnow()
+
+    fuzzer_length = max(len(f) for f in os.listdir(fuzzers))
+
+    logger = Logger(fuzzer, time_start, fuzzer_length)
+
+    def log(msg, *a, **k):
+        logger.log(msg, a, k, flush=True)
+
+    # Need a signal handler on SIGCHLD otherwise get_resource doesn't return
+    # anything.
+    signal.signal(signal.SIGCHLD, lambda sig, frame: None)
+
+    fuzzer_dir = os.path.join(fuzzers, fuzzer)
+    assert os.path.exists(fuzzer_dir), fuzzer_dir
+    # FIXME: The fuzzer output directory should be different from the top level
+    # fuzzer directory.
+    fuzzer_out = fuzzer_dir
+    if not os.path.exists(fuzzer_out):
+        os.makedirs(fuzzer_out)
+    assert os.path.exists(fuzzer_out)
+
+    fuzzer_runok = os.path.join(fuzzer_out, "run.ok")
+
+    if os.path.exists(fuzzer_runok):
+        last_modified = datetime.fromtimestamp(os.stat(fuzzer_runok).st_mtime)
+
+        log(
+            "Skipping as run.ok exists (updated @ {})",
+            last_modified.isoformat())
+    else:
+        log("Starting @ {}", time_start.isoformat())
+
+    log_suffix = ".{}.log".format(time_start.isoformat())
+    fuzzer_stdout = os.path.join(fuzzer_out, "stdout" + log_suffix)
+    fuzzer_stderr = os.path.join(fuzzer_out, "stderr" + log_suffix)
+
+    # Make sure the stdout file exists
+    stdout_fd = open(fuzzer_stdout, "w")
+    stdout_fd.write("Build started {}\n".format(time_start.isoformat()))
+    stdout_fd.write("-" * 75)
+    stdout_fd.write("\n")
+    stdout_fd.flush()
+    os.fsync(stdout_fd)
+    stdout_fd.close()
+
+    stdout_fd = open(fuzzer_stdout, "w+")
+    stdout_fd.seek(0, os.SEEK_END)
+
+    # Make sure the stderr file exists
+    stderr_fd = open(fuzzer_stderr, "w")
+    stderr_fd.write("Build started {}\n".format(time_start.isoformat()))
+    stderr_fd.write("-" * 75)
+    stderr_fd.write("\n")
+    stderr_fd.flush()
+    os.fsync(stderr_fd)
+    stderr_fd.close()
+
+    stderr_fd = open(fuzzer_stderr, "w+")
+    stderr_fd.seek(0, os.SEEK_END)
+
+    p = None
+    try:
+        p = subprocess.Popen(
+            "make -C {} run".format(fuzzer_dir),
+            stdin=subprocess.DEVNULL,
+            stdout=stdout_fd,
+            stderr=stderr_fd,
+            shell=True,
+            close_fds=True)
+
+        while True:
+            try:
+                retcode = p.wait(timeout=10)
+                p = None
+            except subprocess.TimeoutExpired:
+                retcode = None
+
+            if retcode is not None:
```

Comment:
Why not just use Popen.communicate or Popen.wait?

Reply:
You are unable to print out the fact the process is still running. This is an important feature as you have multiple fuzzers running at a time you want to understand which ones are still running.